### ISSUE TITLE
Custom GRU implementation

### ISSUE THREAD

**Issue by sytham** [openai/rllab#138]
Hi, out of curiosity -- and for implementing my own stuff -- I was wondering why rllab has its own implementation of GRU, and doesn't use the one in lasagne? Is that because the lasagne "stateful" implementation (i.e. ability to compute output per time step, remembering previous hidden state) isn't working / satisfactory?

**Comment by dementrock** [comment]
I didn't know about the stateful implementation in lasagne. Can you point me to that?

**Comment by sytham** [comment]
Well, as far as I understand from [this recipe](https://github.com/Lasagne/Recipes/pull/5/files) you can use the hid_init keyword of `GRULayer` to maintain state, i.e. you pass the previous hidden state to hid_init on every batch update.

Then again if I understand [this thread](https://groups.google.com/forum/#!topic/lasagne-users/9jF7c49oMoI) correctly this is not as easy to do for LSTMs, and the ability to do that for LSTMs is not part of Lasagne master yet.

I'm still trying to get my head around why all these complications are necessary anyway, but I guess it's because the computation graph has to be a DAG?

**Comment by dementrock** [comment]
Sorry for my late reply! This seems like a potentially feasible approach. You are welcome to make a pull request of a recurrent policy implementation utilizing this interface.

**Comment by sytham** [comment]
No worries. I tried to do something like that, it worked in general but not inside rllab, I think because of some issues with the likelihood calculation, don't remember exactly. In any case I had no time to look into it further so I dropped it and used the current rllab mechanism for implementing other RNNs.

So back to my original question, it is the case then that rllab has its own implementation in order to have "stateful" RNNs, and not because of some other factors?
